File: src/java/magma/list/RangeHead.java
package magma.list;

import magma.option.None;
import magma.option.Option;
import magma.option.Some;

public final class RangeHead implements Head<Integer> {
    private final int extent;
    private int counter = 0;

    public RangeHead(int extent) {
        this.extent = extent;
    }

    @Override
    public Option<Integer> next() {
        if (counter >= extent) return new None<>();
        int current = counter;
        counter++;
        return new Some<>(current);
    }
}

File: src/windows/magma/api/collect/head/RangeHead.c
#include "./RangeHead.h"
export class RangeHead implements Head<number> {
	length: number;
	mut counter: number;
}

constructor (length: number) {
	this.length = length;
	this.counter = 0;
}
mut next(): Option<number> {
	if (this.counter >= this.length) {
		return new None<number>();
	}
	let value = this.counter;
	this.counter++;
	return new Some<number>(value);
}